connection: expose setters for STARTUP options

wprzytula opened the pull request:
Resolves: #1028

## Motivation
See #1028.

# What's done

## Preparatory steps

### Reduce allocations by using `Cow<str>` for SUPPORTED/STARTUP options
- `types::write_string_map` was genericised to accept `HashMap`s with keys and values being any form of strings, including `Cow<str>`.
- as majority of keys and significant number of values of STARTUP options are known at compile time and thus statically allocated, `Startup` request now holds `HashMap<Cow<str>, Cow<str>`  instead of `HashMap<String, String>` to reduce needless allocations.

### refactors for readability
- `setup_tcp_keepalive()` was extracted from `Connection::new()` to reduce clutter.
- keys of SUPPORTED/STARTUP options were extracted as constants to `options` module to have them in one place and avoid using plain string literals in code.
- resolved a dated TODO about whether logic in `open_named_connection` should be part of `Connection::new` - answered the question, justified the answer and added docstrings to convey that answer.
- added many comments to `open_named_connection`, as the function is big and crucial for driver operation, so future readers will benefit.

### Compression-related small enhancements
- `Compression` name is now no longer allocated, as its name is a `&'static str` anyway.
- warning is now emitted if requested Compression is not supported; before, it would silently fall back to no compression.

## `SelfIdentity` struct and API
- grouped self-identity-related STARTUP options in `SelfIdentity` struct. It includes:
  - `DRIVER_NAME`
  - `DRIVER_VERSION`
  - `APPLICATION_NAME`
  - `APPLICATION_VERSION`
  - `CLIENT_ID`
  
  `DRIVER_NAME` and `DRIVER_VERSION` are set by default to _"ScyllaDB Rust Driver"_ (previously _"scylla-rust-driver"_, changed to be consistent with _"ScyllaDB Python Driver"_) and Rust driver's package version as declared by Cargo. The remaining ones are unset by default.

- exposed `SelfIdentity`'s API in `SessionBuilder`, and added a docstring with an example usage.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

muzarski reviewed (changes requested):
Good job on adding comments to `connection` module. I always feel like I'm lost trying to understand the logic that resides there.

The commit message mentions:
```
Settable identity include:
- DRIVER_NAME
- DRIVER_VERSION
- APPLICATION_NAME
- APPLICATION_VERSION
- CLIENT_ID

All five are visible in Cassandra's `system_views.clients` in
`client_options` column.

DRIVER_NAME & DRIVER_VERSION are visible in ScyllaDB's `system.clients`.
```

Maybe we should add some test cases that verify it? Especially for the newly introduced STARTUP options.

Lorak-mmk commented:
Where does `APPLICATION_NAME` and `APPLICATION_VERSION` come from?
PR description reads like those attributes are as supported and known as `DRIVER_NAME` / `DRIVER_VERSION`, but:
- there are no columns in system.clients dedicated to those attributes
-  `APPLICATION_NAME` / `APPLICATION_VERSION` are not present in Scylla's source code at all
- only 2 results in Cassandra source is some example in docs: https://github.com/apache/cassandra/blob/d3cbf9c1f72057d2a5da9df8ed567d20cd272931/doc/modules/cassandra/pages/managing/operating/virtualtables.adoc?plain=1#L218 . `APPLICATION_NAME` and `APPLICATION_VERSION` appears in `client_options` which IIUC is an arbitrary dict where client can send any keys.
- driver variables are mentioned in protocol v5 (https://github.com/apache/cassandra/blob/d3cbf9c1f72057d2a5da9df8ed567d20cd272931/doc/native_protocol_v5.spec#L480 ), application variables are not.

Please explain in the PR description (and maybe commit message too):
- Why support those specific keys instead of arbitrary keys
- The difference in support between driver variables and application variables
- Where those application variables come from.

Lorak-mmk commented on scylla-cql/src/frame/request/options.rs:
```diff
@@ -11,3 +11,19 @@ impl SerializableRequest for Options {
+/* Value names for options in SUPPORTED/STARTUP */
+pub const DEFAULT_CQL_PROTOCOL_VERSION: &str = "4.0.0";
+pub const DEFAULT_DRIVER_NAME: &str = "scylla-rust-driver";
+// TODO: consider using `env!`, which raises compile-time error if constant is not defined.
+// I think the constant should be always defined, no need to allow for it to be undefined.
```
Docs agree with you: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates
If the version is not in the manifest then the variable will be set to an empty string. I think you can use `env!`

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -1617,6 +1617,10 @@ pub(crate) async fn open_named_connection(
             );
         } else {
             // Fall back to no compression
+            tracing::warn!(
+                "Requested compression <{}> is not supported by the cluster. Falling back to no compression",
+                compression_str
+            );
```
when can this happen?

wprzytula replied:
This may happen if SUPPORTED message sent by a node does not contain the particular compression string in the value for the `"COMPRESSION"` key.

Lorak-mmk commented on scylla/src/transport/session_builder.rs:
```diff
@@ -942,6 +943,37 @@ impl<K: SessionBuilderKind> GenericSessionBuilder<K> {
         self
     }
+
+    /// Set the custom identity of the driver/application/instance,
+    /// to be sent as options in STARTUP message.
+    ///
+    /// By default driver name and version are sent;
+    /// application name and version and client id are not sent.
+    /// # Example
```
nit: maybe add a newline before `# Example`?

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -1515,6 +1517,11 @@ async fn maybe_translated_addr(
+/// - registers for all event types using REGISTER request (if this is control connection).
+///
+/// At the beginning, translates node's address, if it is subject to address translation.
 pub(crate) async fn open_connection(
     endpoint: UntranslatedEndpoint,
     source_port: Option<u16>,
```
I know that `open_named_connection` is later removed and the name is taken from identity, so this is not important, but: open_connection always sets name to `scylla-rust-driver` and it is the only caller of `open_named_connection`, so driver name will always be `scylla-rust-driver` - however previous commit claimed to change default name. How is it changed if it is never used? Idk, maybe I'm reading the code incorrectly.

wprzytula replied:
That's a correct observation. I'll adjust the code correspondingly.

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -1578,22 +1583,29 @@ pub(crate) async fn open_named_connection(
+    let mut options = HashMap::new();
+    protocol_features.add_startup_options(&mut options);
+
+    // The only CQL protocol version supported by the driver.
```
This change is not only adding comments. But this reordering seems harmelss, so let's ignore it.

wprzytula commented:
Rebased on main.

wprzytula commented:
> Where does `APPLICATION_NAME` and `APPLICATION_VERSION` come from? PR description reads like those attributes are as supported and known as `DRIVER_NAME` / `DRIVER_VERSION`, but:
> 
>     * there are no columns in system.clients dedicated to those attributes
> 
>     * `APPLICATION_NAME` / `APPLICATION_VERSION` are not present in Scylla's source code at all
> 
>     * only 2 results in Cassandra source is some example in docs: https://github.com/apache/cassandra/blob/d3cbf9c1f72057d2a5da9df8ed567d20cd272931/doc/modules/cassandra/pages/managing/operating/virtualtables.adoc?plain=1#L218 . `APPLICATION_NAME` and `APPLICATION_VERSION` appears in `client_options` which IIUC is an arbitrary dict where client can send any keys.
> 
>     * driver variables are mentioned in protocol v5 (https://github.com/apache/cassandra/blob/d3cbf9c1f72057d2a5da9df8ed567d20cd272931/doc/native_protocol_v5.spec#L480 ), application variables are not.
> 
> 
> Please explain in the PR description (and maybe commit message too):
> 
>     * Why support those specific keys instead of arbitrary keys
> 
>     * The difference in support between driver variables and application variables
> 
>     * Where those application variables come from.

They come directly from CPP driver. See https://github.com/scylladb/cpp-driver/blob/fa0f27069a625057984d1fa58f434ea99b86c83f/include/cassandra.h#L2916.
As we want to support as big subset of its API as possible in cpp-rust-driver, I decided to expose API for setting those particular key-value pairs, similarly to what cpp-driver does, and not an API to set arbitrary key-value pairs.

Allowing users to set arbitrary options could break the driver by overwriting options that bear special meaning, e.g. the shard-aware port. Therefore, I'm against such liberal API. OTOH, we need to expose APPLICATION_NAME, APPLICATION_VERSION and CLIENT_ID for cpp-rust-driver.

Does this convince you @Lorak-mmk?

Lorak-mmk commented:
> > Where does `APPLICATION_NAME` and `APPLICATION_VERSION` come from? PR description reads like those attributes are as supported and known as `DRIVER_NAME` / `DRIVER_VERSION`, but:
> > ```
> > * there are no columns in system.clients dedicated to those attributes
> > 
> > * `APPLICATION_NAME` / `APPLICATION_VERSION` are not present in Scylla's source code at all
> > 
> > * only 2 results in Cassandra source is some example in docs: https://github.com/apache/cassandra/blob/d3cbf9c1f72057d2a5da9df8ed567d20cd272931/doc/modules/cassandra/pages/managing/operating/virtualtables.adoc?plain=1#L218 . `APPLICATION_NAME` and `APPLICATION_VERSION` appears in `client_options` which IIUC is an arbitrary dict where client can send any keys.
> > 
> > * driver variables are mentioned in protocol v5 (https://github.com/apache/cassandra/blob/d3cbf9c1f72057d2a5da9df8ed567d20cd272931/doc/native_protocol_v5.spec#L480 ), application variables are not.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please explain in the PR description (and maybe commit message too):
> > ```
> > * Why support those specific keys instead of arbitrary keys
> > 
> > * The difference in support between driver variables and application variables
> > 
> > * Where those application variables come from.
> > ```
> 
> They come directly from CPP driver. See https://github.com/scylladb/cpp-driver/blob/fa0f27069a625057984d1fa58f434ea99b86c83f/include/cassandra.h#L2916. As we want to support as big subset of its API as possible in cpp-rust-driver, I decided to expose API for setting those particular key-value pairs, similarly to what cpp-driver does, and not an API to set arbitrary key-value pairs.
> 
> Allowing users to set arbitrary options could break the driver by overwriting options that bear special meaning, e.g. the shard-aware port. Therefore, I'm against such liberal API. OTOH, we need to expose APPLICATION_NAME, APPLICATION_VERSION and CLIENT_ID for cpp-rust-driver.
> 
> Does this convince you @Lorak-mmk?

It does. Please include this explanation in commit message and in a comment.

wprzytula commented:
v2:
- addressed review suggestions;
- made `SelfIdentity` more convenient by accepting `impl Into<Cow<...>>` instead of only `Cow<...>`;
- wrote a test using proxy that asserts proper self identity being set in STARTUP frame.
